src/gates/public_input.rs

```diff
@@ -64,35 +64,36 @@ impl<C: HaloCurve> Gate<C> for PublicInputGate<C> {
 
 impl<C: HaloCurve> WitnessGenerator<C::ScalarField> for PublicInputGate<C> {
     fn dependencies(&self) -> Vec<Target<C::ScalarField>> {
-        (0..NUM_WIRES)
-            .map(|i| {
-                Target::Wire(Wire {
-                    gate: self.index,
-                    input: i,
-                })
-            })
-            .collect()
+        Vec::new()
     }
 
     fn generate(
         &self,
         _constants: &[Vec<C::ScalarField>],
         witness: &PartialWitness<C::ScalarField>,
     ) -> PartialWitness<C::ScalarField> {
-        let self_as_generator: &dyn WitnessGenerator<C::ScalarField> = self;
-        let targets: Vec<Target<C::ScalarField>> = self_as_generator.dependencies();
+        let targets: Vec<Target<C::ScalarField>> = (0..NUM_WIRES)
+            .map(|i| {
+                Target::Wire(Wire {
+                    gate: self.index,
+                    input: i,
+                })
+            })
+            .collect();
 
         let mut result = PartialWitness::new();
         for i_advice in 0..NUM_ADVICE_WIRES {
             let i_wire = NUM_ROUTED_WIRES + i_advice;
-            let value = witness.get_target(targets[i_wire]);
-            result.set_wire(
-                Wire {
-                    gate: self.index + 1,
-                    input: i_advice,
-                },
-                value,
-            );
+            if witness.contains_target(targets[i_wire]) {
+                let value = witness.get_target(targets[i_wire]);
+                result.set_wire(
+                    Wire {
+                        gate: self.index + 1,
+                        input: i_advice,
+                    },
+                    value,
+                );
+            }
```

Comment:
It turns out this was broken since `generate()` wasn't called unless all wires of the public input gate were set. The code worked because of how `PartialWitness::set_public_input` worked. Since I've removed this function, the bug here appeared. 
The fix is rather hacky, but I didn't see a better way to do this without modifying the `WitnessGenerator` trait.

Reply:
Ah yeah I see the problem with the old code, but I think this approach wouldn't work in all cases. I.e. since the generator doesn't declare any deps, it may be run "too early" and no-op. (The code that drives generators only executes each generator once, so there wouldn't be another chance for the generator to run again later.)

It seems like (unless we change the generator model) there's just no way for this to work with a single generator for a whole `PublicInputGate`. Some possible workarounds:
- We could change the gate model so that a gate can register multiple generators, rather than a gate "being" a generator. This seems pretty natural to me, though it might make the gate code somewhat verbose. Relatedly, this would mean that gates don't necessarily need to store their indices in the witness, which might make circuit definitions slightly cleaner.
- We could have `set_target` handle public input targets in a special way, adding two entries to the witness, like `set_public_input` did. Might need some other changes to make sure everything goes through `set_target`; I'd need to double check.

What do you think?

Reply:
Or on second thought, does `PublicInputGate` still need a generator at all? I was thinking `PartialWitness` could store only the new `PublicInput` targets, and the two associated `Wire`s wouldn't appear at all until the `PartialWitness` was converted to a `Witness`.

Reply:
Oh yeah you're right. I didn't take into account the cases where some `Target::PublicInput(_)` haven't been set before calling `Circuit::generate_witness`. I've added a test that was failing with the previous version (before commit 3db3dae). 

I've added a fix that looks in the witness for targets in buffer gates following a public input gate. And if some are found, adds the corresponding non-routable wire in the public input gate to the partial witness. 
It's in `PartialWitnesss::copy_buffer_to_pi_gate` and is called after each call to `Circuit::generate_copies` in the witness generation.
I think it's easier to do it this way, and I couldn't find a way to make it work by modifying `set_target` without having to change a lot of existing code.

> We could change the gate model so that a gate can register multiple generators, rather than a gate "being" a generator. This seems pretty natural to me, though it might make the gate code somewhat verbose. Relatedly, this would mean that gates don't necessarily need to store their indices in the witness, which might make circuit definitions slightly cleaner.

I think this would be a good idea, but would require a relatively large refactor. Maybe we can keep it for when we rewrite Plonky.

Reply:
Hm I think there may still be a couple troublesome cases, maybe we could talk through it tomorrow. Also it looks like `copy_buffer_to_pi_gate` would make generation quadratic-time, which could be OK as a temporary fix but I think we should aim for quasi-linear long-term. (Currently I think it's linear under certain assumptions.)

It seems tricky to get all the cases right since public inputs have three possible forms in `PartialWitness` (`Target::PublicInput`, original wire, routable wire). I think it may be easiest if we only have `Target::PublicInput` in `PartialWitness` (and hide `original_wire`/`routable_target` so that the user can't interact with the other forms), and introduce the other forms when converting `PartialWitness` to `Witness` and `TargetPartitions` to `WirePartitions`.

Reply:
Hmm I'm not sure that the complexity changes. `copy_buffer_to_pi(copy_results)` runs in time linear in `|copy_results|` which is also the same as `witness.extend(copy_results)`.
But I agree that this is maybe not the best design. We can talk about it more tomorrow.

Reply:
Ah yeah I misread that; I thought `copy_buffer_to_pi` was iterating through the "main" `PartialWitness`.